platforms/software/dependency-management/src/main/java/org/gradle/internal/resolve/resolver/ExcludingVariantArtifactSet.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.resolve.resolver;
+
+import org.gradle.api.artifacts.ModuleIdentifier;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvableArtifact;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvedArtifactSet;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvedVariant;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.excludes.specs.ExcludeSpec;
+import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.internal.Describables;
+import org.gradle.internal.DisplayName;
+import org.gradle.internal.component.external.model.ImmutableCapabilities;
+import org.gradle.internal.component.model.VariantResolveMetadata;
+
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+/**
+ * A {@link ResolvedVariant} that applies artifact exclusions to a delegate {@link ResolvedVariant}.
+ */
+public class ExcludingVariantArtifactSet implements ResolvedVariant, VariantResolveMetadata.Identifier {
+
+    private final ResolvedVariant delegate;
+    private final ModuleIdentifier moduleId;
+    private final ExcludeSpec exclusions;
```

Comment:
💭 Another type that probably needs refactoring and/or renaming.  Seems like there should be separate `IvyModuleExcludeSpec` and `ArtifactExcludeSpec` (to be used here).

Reply:
We should have separate code dealing with module exclusions, which can be applied when building the graph, and file exclusions which are applied during artifact resolution.

One problem with this is we put in a lot of work to merge exclusions recursively in graph resolution. If we separate these two types, we need to duplicate that code or somehow have them wrapped by some common interface